### PR TITLE
Add nexthop strategy plugins

### PR DESCRIPTION
**Pull request by rob05c**
Adds a remap plugin hook to add a nexthop strategy object, and moves and adds other necessary API symbols necessary to implement strategies in plugins.

Also adds an example nexthop strategy plugin.

This adds the following to the API:
```
enum TSHostStatus
enum TSHostStatusReason
enum TSParentResultType
class TSNextHopSelectionStrategy
struct TSParentResult
const char *ParentResultStr
bool TSHostnameIsSelf(const char* hostname)
bool TSHostStatusGet(const char *hostname, TSHostStatus* status, unsigned int *reason)
void TSHostStatusSet(const char *hostname, TSHostStatus status, const unsigned int down_time, const unsigned int reason)
TSParentResult* TSHttpTxnParentResultGet(TSHttpTxn txnp)
void TSHttpTxnParentResultSet(TSHttpTxn txnp, struct TSParentResult* result)
TSReturnCode TSRemapInitStrategy(TSNextHopSelectionStrategy *&strategy, void *ih, char *errbuf, int errbuf_size)
```

I tried to minimize API changes, but I didn't see a way to do it any smaller. 
The example plugin is also a lot of duplication from the core ConsistentHash strategy. I didn't see a reasonable way to reduce that either, while keeping the ConsistentHash strategy in core for now. 
I'm definitely open to suggestions.

### CODE REVIEW

**Comment by randall** [comment]
[approve ci freebsd]

File: include/ts/ts.h [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
 #pragma once
 
+#include <string.h>
```
Comment (SolidWallOfCode) [inline include/ts/ts.h:32]: Why is this needed?

File: include/ts/nexthop.h [hunk 2]
```diff
@@ -30,20 +30,20 @@
...
-// plugin callback commands.
 enum NHCmd { NH_MARK_UP, NH_MARK_DOWN };
 
-struct NHHealthStatus {
-  virtual bool isNextHopAvailable(TSHttpTxn txn, const char *hostname, const int port, void *ih = nullptr) = 0;
-  virtual void markNextHop(TSHttpTxn txn, const char *hostname, const int port, const NHCmd status, void *ih = nullptr,
-                           const time_t now = 0)                                                           = 0;
-  virtual ~NHHealthStatus() {}
-};
+class TSNextHopSelectionStrategy
```
Comment (SolidWallOfCode) [inline include/ts/nexthop.h:35]: Does this make the plugin API require C++?

File: include/ts/parentresult.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+#pragma once
+
+#include "ts/apidefs.h"
+#include "tscore/ConsistentHash.h"
+
+// Needed by core. Disabling the warning for plugins
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-variable"
+static const char *ParentResultStr[] = {"PARENT_UNDEFINED", "PARENT_DIRECT", "PARENT_SPECIFIED", "PARENT_AGENT", "PARENT_FAIL"};
```
Comment (SolidWallOfCode) [inline include/ts/parentresult.h:30]: This should be defined in a ".cc" file and only the extern should be here. That should (1) avoid having multiple copies of this table and (2) avoid the warning.

File: include/ts/parentresult.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#pragma GCC diagnostic pop
+
+struct TSParentResult {
+  const char *hostname;
+  int port;
+  bool retry;
+  TSParentResultType result;
+  bool chash_init[TS_MAX_GROUP_RINGS] = {false};
```
Comment (SolidWallOfCode) [inline include/ts/parentresult.h:38]: Hmmm, is this C++? If so, `std::bitset` is a better choice here, otherwise I'd look at making this a bit mask object.

File: include/ts/ts.h (hunk 1 above)
Comment (rob05c) [reply include/ts/ts.h]: It's not. Removed

File: include/ts/nexthop.h (hunk 2 above)
Comment (rob05c) [reply include/ts/nexthop.h:35]: I can move this to `include/tscpp`. Is that ok?

Sort of. `TSNextHopSelectionStrategy` is used by the Remap hook `TSRemapInitStrategy`, not in InkApi.cc. A Remap plugin can still be C and just not implement that function.

Likewise, `TSHttpTxnParentResult(G|S)et` in InkApi.cc uses a `TSParentResult`, which has C++ stuff in it. I just exposed that struct to the API, I didn't create it. If someone doesn't need to call that function, if I move `parentresult.h` to `include/tscpp`, someone can write a C plugin and just not call the functions that use it.

Is that ok? Or is there a better way?

File: include/ts/parentresult.h (hunk 3 above)
Comment (rob05c) [reply include/ts/parentresult.h]: Done

File: include/ts/parentresult.h (hunk 4 above)
Comment (rob05c) [reply include/ts/parentresult.h]: Yeah, the object is C++, unfortunately.

This is an API version of a core struct that already exists. If it's changed here, moving the data between the API and core will have to translate between them, and will probably be less efficient.

Alternatively, I could change it in core. It looks like that wouldn't be too invasive.

I haven't benchmarked, but it was my understanding that std::bitset is optimized for space, and is often less performant than `bool[]` or bitmasking. Notwithstanding, I'd still prefer to do a bitmask, because the struct only has one C++ member, and I'd like to move it in the direction of C for the sake of the API.

File: include/ts/ts.h [hunk 5]
```diff
@@ -2573,6 +2573,38 @@ tsapi TSIOBufferReader TSHttpTxnPostBufferReaderGet(TSHttpTxn txnp);
...
 
+/*
+ * Returns TS_SUCCESS if hostname is this machine, as used for parent and remap self-detection.
+ * Returns TS_ERROR if hostname is not this machine.
+ */
+tsapi TSReturnCode TSHostnameIsSelf(const char *hostname);
+
+/*
+ * Gets the status of hostname in the outparam status, and the status reason in the outparam reason.
+ * The reason is a logical-and combination of the reasons in TSHostStatusReason.
```
Comment (SolidWallOfCode) [inline include/ts/ts.h:2584]: Logical or, maybe? Logical and would always be zero.

File: include/tscpp/api/parentresult.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#pragma once
+
+#include "ts/apidefs.h"
+#include "tscore/ConsistentHash.h"
+
+extern const char *ParentResultStr[];
+
+struct TSParentResult {
+  const char *hostname;
+  int port;
```
Comment (SolidWallOfCode) [inline include/tscpp/api/parentresult.h:31]: `in_port_t`

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+// hash_key strings.
+constexpr std::string_view hash_key_url           = "url";
+constexpr std::string_view hash_key_hostname      = "hostname";
+constexpr std::string_view hash_key_path          = "path";
+constexpr std::string_view hash_key_path_query    = "path+query";
+constexpr std::string_view hash_key_path_fragment = "path+fragment";
+constexpr std::string_view hash_key_cache         = "cache_key";
+
+static HostRecord *
+chash_lookup(std::shared_ptr<ATSConsistentHash> ring, uint64_t hash_key, ATSConsistentHashIter *iter, bool *wrapped,
```
Comment (SolidWallOfCode) [inline plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:55]: This is more than enough state to justify an iterator.

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+  HostRecord *host_rec = nullptr;
+
+  if (*hash_init == false) {
+    host_rec   = static_cast<HostRecord *>(ring->lookup_by_hashval(hash_key, iter, wrapped));
+    *hash_init = true;
+  } else {
+    host_rec = static_cast<HostRecord *>(ring->lookup(nullptr, iter, wrapped, hash));
+  }
+  bool wrap_around = *wrapped;
+  *wrapped         = (*mapWrapped && *wrapped) ? true : false;
```
Comment (SolidWallOfCode) [inline plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:67]: Why the ternary? Just `*wrapped = (*mapWrapped && *wrapped) ;`

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+
+  // load up the hash rings.
+  for (uint32_t i = 0; i < groups; i++) {
+    std::shared_ptr<ATSConsistentHash> hash_ring = std::make_shared<ATSConsistentHash>();
+    for (uint32_t j = 0; j < host_groups[i].size(); j++) {
+      // ATSConsistentHash needs the raw pointer.
+      HostRecord *p = host_groups[i][j].get();
+      // need to copy the 'hash_string' or 'hostname' cstring to 'name' for insertion into ATSConsistentHash.
+      if (!p->hash_string.empty()) {
+        p->name = const_cast<char *>(p->hash_string.c_str());
```
Comment (SolidWallOfCode) [inline plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:126]: Is `const_cast` really required here? It's a bit risky.

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+        p->name = const_cast<char *>(p->hostname.c_str());
+      }
+      p->group_index = host_groups[i][j]->group_index;
+      p->host_index  = host_groups[i][j]->host_index;
+      hash_ring->insert(p, p->weight, &hash);
+      NH_Debug(NH_DEBUG_TAG, "Loading hash rings - ring: %d, host record: %d, name: %s, hostname: %s, stategy: %s", i, j, p->name,
+               p->hostname.c_str(), strategy_name.c_str());
+    }
+    hash.clear();
+    rings.push_back(std::move(hash_ring));
```
Comment (SolidWallOfCode) [inline plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:137]: Don't `move` a `shared_ptr`.

File: proxy/ParentSelection.cc [hunk 11]
```diff
@@ -1869,23 +1871,65 @@ void
...
+  this->wrap_around         = r->wrap_around;
+  memcpy(this->mapWrapped, r->mapWrapped, 2 * sizeof(bool));
+  this->last_lookup = r->last_lookup;
+  for (int i = 0; i < TS_MAX_GROUP_RINGS; ++i) {
+    this->chashIter[i] = r->chashIter[i];
+  }
+}
+
+void
+ParentResult::copyTo(TSParentResult *r)
```
Comment (SolidWallOfCode) [inline proxy/ParentSelection.cc:1917]: Any reason these aren't assignment operators?

File: src/traffic_server/InkAPI.cc [hunk 12]
```diff
@@ -9790,6 +9791,54 @@ TSHttpTxnRedoCacheLookup(TSHttpTxn txnp, const char *url, int length)
   return TS_ERROR;
 }
 
+TSReturnCode
+TSHostnameIsSelf(const char *hostname)
+{
+  const bool isSelf = Machine::instance()->is_self(hostname);
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc:9797]:
```
return Machine::instance()->is_self(hostname) ? TS_SUCCESS : TS_ERROR;
```
?

File: src/traffic_server/InkAPI.cc [hunk 13]
```diff
@@ -9790,6 +9791,54 @@ TSHttpTxnRedoCacheLookup(TSHttpTxn txnp, const char *url, int length)
...
+  if (isSelf) {
+    return TS_SUCCESS;
+  }
+  return TS_ERROR;
+}
+
+TSReturnCode
+TSHostStatusGet(const char *hostname, TSHostStatus *status, unsigned int *reason)
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc:9805]: As a general rule, I prefer the C API functions to take strings as `const char *, int` which is commonly done in (for instance) the MIME header functions. This makes working with `string_view` (or `TextView`) in plugins much easier. Internally `HostStatus::getHostStatus` should take an `string_view` parameter so that it can be directly passed from this API call.

File: proxy/http/remap/RemapPluginInfo.cc [hunk 14]
```diff
@@ -271,6 +273,31 @@ RemapPluginInfo::indicatePostReload(TSRemapReloadStatus reloadStatus)
...
+RemapPluginInfo::initStrategy(void *ih)
+{
+  if (!init_strategy_cb) {
+    PluginDebug(_tag, "plugin '%s' has no init_strategy_cb, returning nullptr", _configPath.c_str());
+    return std::shared_ptr<TSNextHopSelectionStrategy>();
+  }
+  char tmpbuf[2048];
+  ink_zero(tmpbuf);
+  TSNextHopSelectionStrategy *strategy_raw = nullptr;
+  if (init_strategy_cb(strategy_raw, ih, tmpbuf, sizeof(tmpbuf) - 1) != TS_SUCCESS) {
```
Comment (SolidWallOfCode) [inline proxy/http/remap/RemapPluginInfo.cc:287]: How does `strategy_raw` become not `nullptr`?

File: proxy/http/remap/PluginFactory.h [hunk 15]
```diff
@@ -54,6 +54,10 @@ class RemapPluginInst
   TSRemapStatus doRemap(TSHttpTxn rh, TSRemapRequestInfo *rri);
   void osResponse(TSHttpTxn rh, int os_response_type);
 
+  /* gets the Strategy if the plugin has a TSRemapInitStrategy function. Returns a nullptr if not. */
+  std::shared_ptr<TSNextHopSelectionStrategy> getStrategy();
+  std::string name() const;
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginFactory.h:59]: Noooooooooo! Don't return a `std::string`.

**Comment by a-canary** [comment]
-0, AFAIK the general plugin design is to use event handlers. Which I realized will take a lot more work but also simplify the API and the HttpSM quite a bit in the end. I hope we can move towards that once Susan gets the ConnectionSM refactored.

File: include/tscpp/api/nexthop.h [hunk 16]
```diff
@@ -30,20 +30,20 @@
...
+  TSNextHopSelectionStrategy(){};
+  virtual ~TSNextHopSelectionStrategy(){};
 
-struct NHPluginStrategy {
-  virtual void findNextHop(TSHttpTxn txnp, void *ih = nullptr)   = 0;
-  virtual bool nextHopExists(TSHttpTxn txnp, void *ih = nullptr) = 0;
-  virtual ~NHPluginStrategy() {}
+  virtual void findNextHop(TSHttpTxn txnp, time_t now = 0)                                                                 = 0;
+  virtual void markNextHop(TSHttpTxn txnp, const char *hostname, const int port, const NHCmd status, const time_t now = 0) = 0;
+  virtual bool nextHopExists(TSHttpTxn txnp)                                                                               = 0;
```
Comment (SolidWallOfCode) [inline include/tscpp/api/nexthop.h:43]: Is this really useful? Why not have `findNextHop` return `bool`?

File: include/tscpp/api/parentresult.h [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+  bool chash_init[TS_MAX_GROUP_RINGS] = {false};
+  TSHostStatus first_choice_status    = TSHostStatus::TS_HOST_STATUS_INIT;
+  int line_number;
+  uint32_t last_parent;
+  uint32_t start_parent;
+  uint32_t last_group;
+  bool wrap_around;
+  bool mapWrapped[2];
+  int last_lookup;
+  ATSConsistentHashIter chashIter[TS_MAX_GROUP_RINGS];
```
Comment (SolidWallOfCode) [inline include/tscpp/api/parentresult.h:43]: Are the hash values and iterators needed in this result structure? Is this intended as a state object and not just a result?

File: proxy/http/remap/PluginFactory.cc [hunk 18]
```diff
@@ -76,6 +76,18 @@ RemapPluginInst::doRemap(TSHttpTxn rh, TSRemapRequestInfo *rri)
   return _plugin.doRemap(_instance, rh, rri);
 }
 
+std::shared_ptr<TSNextHopSelectionStrategy>
+RemapPluginInst::getStrategy()
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginFactory.cc:80]: There seems a bit of a name semantic mismatch - the outer is "get" but the internal is "init". Does this initialize or just get?

File: include/ts/ts.h (hunk 5 above)
Comment (rob05c) [reply include/ts/ts.h]: This was moved from core. Fixed

File: include/tscpp/api/parentresult.h (hunk 6 above)
Comment (rob05c) [reply include/tscpp/api/parentresult.h]: Copied from core. Fixed

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc (hunk 8 above)
Comment (rob05c) [reply plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc]: Copied from core. Fixed.
For the record, I would never write `? true : false` 🤦

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc (hunk 9 above)
Comment (rob05c) [reply plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:126]: Also copied from core. Yeah, because `HostRecord.name` is not const. I was told it couldn't be for reasons, and that it would never be changed. Likewise, I was assured hash_string would always outlive it, making returning the c_str safe.

I can try to verify those things, but it would take some time. FWIW I certainly don't like it either, but didn't see an obvious alternative.

File: proxy/ParentSelection.cc (hunk 11 above)
Comment (rob05c) [reply proxy/ParentSelection.cc:1917]: Eh, I lean against obfuscating nontrivial copies, and these funcs are just complex enough to be nontrivial. I can change it if you want / if overriding assignment is the idiom in ATS

File: src/traffic_server/InkAPI.cc (hunk 13 above)
Comment (rob05c) [reply src/traffic_server/InkAPI.cc]: Done

File: proxy/http/remap/RemapPluginInfo.cc (hunk 14 above)
Comment (rob05c) [reply proxy/http/remap/RemapPluginInfo.cc:287]: `init_strategy_cb` is `TSRemapInitStrategy`, which takes a `TSNextHopSelectionStrategy *&`. A plugin sets that pointer-ref. https://github.com/apache/trafficserver/blob/00e236d94aa5577e8a1d0f94d18787bd2c7e85b3/plugins/experimental/nexthop_strategy_consistenthash/plugin_consistenthash.cc#L74

File: proxy/http/remap/PluginFactory.h (hunk 15 above)
Comment (rob05c) [reply proxy/http/remap/PluginFactory.h]: Done

File: include/tscpp/api/nexthop.h (hunk 16 above)
Comment (rob05c) [reply include/tscpp/api/nexthop.h:43]: `findNextHop` modifies the state to be the next parent; `nextHopExists` returns without modifying state.
I'd be a big +1 on making `findNextHop` return and be a pure function. But that'd require refactoring how Strategies work and are called in core. Which I'm guessing itself is a reflection of how ParentSelection always did it.

File: include/tscpp/api/parentresult.h (hunk 17 above)
Comment (rob05c) [reply include/tscpp/api/parentresult.h:44]: Yeah, it's a state object. Used by both ParentSelection and Strategies in core (or rather, copied to an identical structure in core).

I.e.
https://github.com/apache/trafficserver/blob/786463/proxy/ParentConsistentHash.cc#L110
https://github.com/apache/trafficserver/blob/786463/proxy/http/remap/NextHopConsistentHash.cc#L335

File: proxy/http/remap/PluginFactory.cc (hunk 18 above)
Comment (rob05c) [reply proxy/http/remap/PluginFactory.cc:80]: This gets the strategy that the func it calls initializes. It's conceptually debatable, I don't object to renaming it

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc (hunk 10 above)
Comment (rob05c) [reply plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc]: Changed

File: proxy/http/remap/PluginFactory.cc (hunk 18 above)
Comment (rob05c) [reply proxy/http/remap/PluginFactory.cc]: Done

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc (hunk 7 above)
Comment (rob05c) [reply plugins/experimental/nexthop_strategy_consistenthash/consistenthash.cc:55]: This was moved/copied from core. Not sure what you mean? It takes a `ATSConsistentHashIter`, that's what that is. You mean taking all these args and wrapping them up in an iterator object, so things would call `chIter->begin()`, `++chIter`, etc?

File: plugins/experimental/nexthop_strategy_consistenthash/consistenthash_config.cc [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    strategies = config["strategies"];
+    if (strategies.Type() != YAML::NodeType::Sequence) {
+      NH_Error("malformed %s file, expected a 'strategies' sequence", basename);
+      return nullptr;
+    }
+
+    for (unsigned int i = 0; i < strategies.size(); ++i) {
+      YAML::Node strategy = strategies[i];
```
Comment (brbzull0) [inline plugins/experimental/nexthop_strategy_consistenthash/consistenthash_config.cc:95]: Have you considered following [this](https://github.com/jbeder/yaml-cpp/wiki/Tutorial#converting-tofrom-native-data-types) custom Data types like `TSNextHopSelectionStrategy`?

**Comment by rob05c** [comment]
Changing back to a draft: I have an idea to get the best of both worlds, small core changes like this PR, but also powerful and with real continuations, hooks, etc.

I'm not 100% sure the idea will work. If it does, I'll close those. If not, I'll un-draft it.

**Comment by rob05c** [comment]
Closing in favor of https://github.com/apache/trafficserver/pull/7467
